carapace-server/src/main/java/org/carapaceproxy/cluster/impl/ZooKeeperGroupMembershipHandler.java

```diff
@@ -0,0 +1,167 @@
+import org.carapaceproxy.cluster.GroupMembershipHandler;
+
+/**
+ * Implementation based on ZooKeeper.
+ * This class is very simple, we are not expecting heavy traffic on ZooKeeper.
+ * We have two systems:
+ * <ul>
+ * <li>Peer discovery
+ * <li>Configuration changes event broadcast
+ * </ul>
+ *
+ * @author eolivelli
+ */
+public class ZooKeeperGroupMembershipHandler implements GroupMembershipHandler, AutoCloseable {
+
+    private static final Logger LOG = Logger.getLogger(ZooKeeperGroupMembershipHandler.class.getName());
+
+    private final CuratorFramework client;
+    private final String peerId;
+    private CopyOnWriteArrayList<PathChildrenCache> watchedEvents = new CopyOnWriteArrayList<>();
+
+    public ZooKeeperGroupMembershipHandler(String zkAddress, int zkTimeout, String peerId) {
+        client = CuratorFrameworkFactory
+                .builder()
+                .sessionTimeoutMs(zkTimeout)
+                .waitForShutdownTimeoutMs(1000) // useful for tests
+                .retryPolicy(new ExponentialBackoffRetry(1000, 2))
+                .ensembleProvider(new FixedEnsembleProvider(zkAddress, true))
+                .build();
+        this.peerId = peerId;
+    }
+
+    @Override
+    public void start() {
+        try {
+            client.start();
+            client.blockUntilConnected();
+            {
+                Stat exists = client.checkExists().creatingParentsIfNeeded()
+                        .forPath("/proxy/peers/" + peerId);
+                if (exists == null) {
+                    client.create()
+                            .creatingParentsIfNeeded()
+                            .withMode(CreateMode.EPHEMERAL) // auto delete on close
+                            .forPath("/proxy/peers/" + peerId);
+                }
+            }
+            {
+                final String path = "/proxy/events";
+                Stat exists = client.checkExists().creatingParentsIfNeeded()
+                        .forPath(path);
+                if (exists == null) {
+                    client.create()
+                            .creatingParentsIfNeeded()
+                            .withMode(CreateMode.PERSISTENT)
+                            .forPath(path);
+                }
+            }
+        } catch (Exception ex) {
+            throw new RuntimeException(ex);
+        }
+    }
+
+    @Override
+    public void watchEvent(String eventId, EventCallback callback) {
+        try {
+            final String path = "/proxy/events";
+
+            LOG.info("watching " + path);
+            PathChildrenCache cache = new PathChildrenCache(client, path, true);
+            // hold a strong reference to the PathChildrenCache
+            watchedEvents.add(cache);
+            cache.getListenable().addListener((PathChildrenCacheListener) (CuratorFramework cf, PathChildrenCacheEvent pcce) -> {
+                LOG.log(Level.INFO, "ZK event {0} at {1}", new Object[]{pcce, path});
```

Comment:
this is still work in progress.
watchers are getting notified about every event